### DISCUSSION TITLE
Flows: Allow linking to the same operation many times

### DISCUSSION THREAD

**Discussion by ched-dev** [directus/directus#19675] [Feature Requests]
### Summary

Right now I can only link operations one-to-one, but I would like to link one-to-many to re-use an operation.

### Basic Example

I have a Flow triggered by an update event using several condition operations which could fail. All need to do the same step after - return the existing payload - but I don't want to create multiple of the same final operation. I would prefer to make one operation, and point all three to the single operation.

https://github.com/directus/directus/assets/73146636/ec095876-34d0-416f-8833-3e27b7370569

### Motivation

This will allow users to re-use operations already in the flow. Right now, if you want to run the same operation you have to duplicate it or create a custom operation extension.

### Detailed Design

No new design as I believe the flow chart will handle it.

### Must Have:
- An operations input can have many triggers

### Won't Have:
- A many-to-many operation connection

### Drawbacks

_None Known_

### Alternatives

_None Known_

### Adoption Strategy

I believe a migration would be needed. It will be a new feature.

### Unresolved Questions

Is it worth it to add a many-to-many option to solve this?

**Comment by br41nslug** [comment]
I can absolutely get behind this if it's just merging of lines as described and not splitting of links (which would create parallel execution coming with a ton of complications)

**Comment by rijkvanzanten** [comment]
As long as we can keep the sequential nature of the flow, I'm all about this. I'm a little worried that this can easily cause infinite loops and other parallel problems though. For example, the output of your example transform can feed back into an input that downstream uses the same transform causing 🔁

**Comment by lukas-schaetzle** [comment]
To be fair, it's already possible to do infinite loops with the script operation, so I wouldn't call that a major issue. As a simple security mechanism the flow execution could be canceled after a given timeout.

Also see these 3 former feature requests for the same feature (closed because of too few upvotes at that time): 
- #18791
- #16356
- #17045 (This one is for multiple outputs, though)

**Comment by Nitwel** [comment]
To throw my beef on the grill too, I was multiple times just close to thinking, duck it, I'm gonna add this now while working with flows but then didn't commit to it cause I thaught that maybe there was a design decision made by either @rijkvanzanten or @nickrum to not allow the merge of flow paths.

So I would to hear if there was a reason and if not, I think this definitely should be added to flows as there should be no technical limitation to implement this.

**Reply by rijkvanzanten** [reply]
IIRC the main reason was around the complexity of recognizing unexpected vs intentional recursion.

**Reply by nickrum** [reply]
Yes, that was a deliberate design decision to make the flow of flows easier to understand and (as @rijkvanzanten said), to prevent loops inside a single flow. Our idea was that you could use a sub-flow that is triggered from the main flow if shared logic is needed, but I see how this might be cumbersome.

**Comment by inovosel** [comment]
I'm sooo missing this "merge flow" feature. Let's say you want to use condition flows to check if some relation data was created, updated or deleted and do something accordingly. You can check only one of mentioned things, but on one update relation could have few created and few deleted items at once.
After checking for one condition i want to merge flow so I can check for other conditions as well, but this is currently not possible.

<img width="642" alt="Screenshot 2024-09-24 at 12 13 18" src="https://github.com/user-attachments/assets/acb2ee7a-31ec-4a1d-b455-a7b59a1ea90c">

**Comment by brocococonut** [comment]
This has been in an "under review" state for a little over a year now, any updates on if it got accepted or rejected by the team?

**Reply by rijkvanzanten** [reply]
I'm on the fence. The difficulty with having two entrypoints to the same operation is that there's no guarantee on type signature for the incoming data matches from the two+ previous outputs. For something like a `log` that doesn't matter as there's no output, but if you have two operations that both generate data in a different format it gets super confusing to figure out what the input type is for the following operation

**Reply by brocococonut** [reply]
I've found kind of a workaround, but heck is it a hacky one...
It seems like the main thing stopping it is the database schema having unique indexes on the error and success route IDs, if we ignore the usability of the UI, removing those two indexes seems to allow for multiple bits to go to the same destination. As it is, if you (in the database) manually change an error results and a success result to have the same ID, they show in the interface as leading to the same place and seem to work.
So I'm going to spin up a new instance and try just deleting those indexes and see if that'll at least let me manually set the destinations at the database level. I'll post an update in this thread when I do

**Reply by rijkvanzanten** [reply]
Yup! That'll work. Those unique constraints are there to make sure you never end up with an infinite recursion or a situation where you have unexpected inputs into an operation though. The main concern for me here is the UX footguns rather than technical possibility 🙂

**Comment by mike-grayhat** [comment]
As far as I see right now even the **basic if / else logic** can't be expressed using existing conditional operation. 

Let's say I want to do an action based on a condition, but then I want to merge the control flow back and it's not possible. So this:
```python
if has_new_messages:
  send_notifications()
update_old_records()
ops_1()
ops_2()
...
```
becomes this:
```python
if has_new_messages:
  send_notifications()
  update_old_records()
  ops_1()
  ops_2()
  ...
else:
  update_old_records()
  ops_1()
  ops_2()
  ...
```

**Reply by inovosel** [reply]
More like:
```javascript
if has_new_messages:
  send_notifications()
  update_old_records_v1()
  ops_1_v1()
  ops_2_v1()
  ...
else:
  update_old_records_v2()
  ops_1_v2()
  ops_2_v2()
  ...
```
And in case you have to update some code in ops_2, you have to do identical changes in two places, in ops_2_v1 and ops_2_v2
🤦

**Reply by mike-grayhat** [reply]
I share concerns of the developers that it can be abused to make cyclic dependencies and ruin flows more easily. Maybe in such case it's reasonable to check if existing flow is a DAG, and give user a warning if it's not.

**Reply by inovosel** [reply]
Directus flows already give you easy way to cause "infinite" loops, you can create a flow that triggers itself. It's not really infinite because it will timeout after few seconds so it runs for some 1000 loops.

<img width="1140" alt="Screenshot 2025-04-17 at 23 20 43" src="https://github.com/user-attachments/assets/6caa6787-f5c0-4f9a-abd5-64ec1a6d9356" />

What I want to point out: with great power (of merging flows and reusable code) comes great responsibility. People will always find a way to do something in a way it was not intended.

I would prefer a risk and have flow branching and merging than making incomprehensibly long and repeating flows inside a walled garden.

**Comment by alantiller** [comment]
Any chance this feature has had anymore thought? I need to be able to run a condition to check if something already exists and if not create and if so update then carry on with the rest of the flow.
